openghg/retrieve/_search_helpers.py

```diff
@@ -0,0 +1,138 @@
+"""
+Helper functions for processing search input into form accepted by `Metastore.search`
+"""
+
+import itertools
+from typing import Any, cast, Callable, Optional, TypeVar
+from openghg.util import extract_float
+from openghg.types import Comparable
+
+
+T = TypeVar("T", bound=Comparable)  # types with <=
+
+
+def _in_interval(x: T, start: Optional[T], stop: Optional[T]) -> bool:
+    """Return True if start <= x <= stop; if either start or stop is None, omit
+    the corresponding bound.
+
+    Args:
+        x: value to test for inclusion in the interval [start, stop]
+        start: start of the interval
+        stop: end of the interval
+
+    Returns:
+        True if start <= x <= stop, False otherwise
+    """
+    if start is None and stop is None:
+        return False
+    elif start is None:
+        stop = cast(T, stop)  # to appease mypy
+        return x <= stop
+    elif stop is None:
+        return x >= start
+    else:
+        return start <= x <= stop
+
+
+def _convert_slice_to_test(s: slice, key: Optional[str] = None) -> Callable:
+    """Convert slice to a function that checks if values are in the interval specified by the slice.
+
+    Args:
+        s: slice specifying interval of values
+        key: optional metadata key, used to choose value formatting
+
+    Returns:
+        function that returns True if values are in the slice interval
+    """
+
+    def formatter(x: Any) -> Any:
+        """Formatting if key == 'inlet': try to extract float from strings."""
+        if key == "inlet":
+            if isinstance(x, (int, float)):
+                return x
+            if isinstance(x, str):
+                try:
+                    result = extract_float(x)
+                except ValueError:
+                    return None
+                else:
+                    return result
+            return None
+        return x  # key != "inlet"
+
+    def test_func(x: Any) -> bool:
+        """Return True if start <= formatter(x) <= stop."""
+        return _in_interval(formatter(x), formatter(s.start), formatter(s.stop))  # type: ignore
+
+    return test_func
+
+
+def process_special_queries(search_terms: dict, neg_lookup_flag: Any = "NOT_SET_FLAG") -> dict:
```

Comment:
We could just set the default for `neg_lookup_flag` to be `np.nan` if that's what we decided. We've basically ended up with 2 situations where the metadata key has not been specified (1) where the key is present but the value is "not_set" and (2) where the key isn't present. The `neg_lookup_flag` covers case (2) so we may want to be a bit careful to not conflate the two cases.

Reply:
I suppose we might want to be able to specify this somewhere (possibly in the config?). I can set it to np.nan for now. I just chose something arbitrary because I wasn't sure what we wanted (but didn't want to use `NOT_SET`).

Reply:
Yeah makes sense and I agree we should specify this somewhere. I think that's fine for now though.